src/backend/columnar/columnar_customscan.c

```diff
@@ -491,15 +563,508 @@ RelationIdGetNumberOfAttributes(Oid relationId)
+ * quals. Ordinary quals, such as x = 7, can be pushed down easily. But join
+ * quals of the form "x = y" (where "y" comes from another rel) require the
+ * proper parameterization.
+ *
+ * Paths that require more outer rels can push down more join clauses that
+ * depend on those outer rels. But requiring more outer rels gives the planner
+ * fewer options for the shape of the plan. That means there is a trade-off,
+ * and we should generate plans of various parameterizations, then let the
+ * planner choose. We always need to generate one minimally-parameterized path
+ * (parameterized only by lateral refs, if present) to make sure that at least
+ * one path can be chosen. Then, we generate as many parameterized paths as we
+ * reasonably can.
+ *
+ * The set of all possible parameterizations is the power set of
+ * candidateRelids. The power set has cardinality 2^N, where N is the
+ * cardinality of candidateRelids. To avoid creating a huge number of paths,
+ * limit the depth of the search; the depthLimit is equivalent to the maximum
+ * number of required outer rels (beyond the minimal parameterization) for the
+ * path. A depthLimit of zero means that only the minimally-parameterized path
+ * will be generated.
+ */
+static void
+AddColumnarScanPathsRec(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte,
+						Relids paramRelids, Relids candidateRelids,
+						int depthLimit)
+{
+	CHECK_FOR_INTERRUPTS();
+	check_stack_depth();
+
+	Assert(!bms_overlap(paramRelids, candidateRelids));
+	AddColumnarScanPath(root, rel, rte, paramRelids);
+
+	/* recurse for all candidateRelids, unless we hit the depth limit */
+	Assert(depthLimit >= 0);
+	if (depthLimit-- == 0)
+	{
+		return;
+	}
+
+	/*
+	 * Iterate through parameter combinations depth-first. Deeper levels
+	 * generate paths of greater parameterization (and hopefully lower
+	 * cost).
+	 */
+	Relids tmpCandidateRelids = bms_copy(candidateRelids);
+	int relid = -1;
+	while ((relid = bms_next_member(candidateRelids, relid)) >= 0)
+	{
+		Relids tmpParamRelids = bms_add_member(
+			bms_copy(paramRelids), relid);
+
+		/*
+		 * Because we are generating combinations (not permutations), remove
+		 * the relid from the set of candidates at this level as we descend to
+		 * the next.
+		 */
+		tmpCandidateRelids = bms_del_member(tmpCandidateRelids, relid);
+
+		AddColumnarScanPathsRec(root, rel, rte, tmpParamRelids,
+								tmpCandidateRelids, depthLimit);
+	}
+
+	bms_free(tmpCandidateRelids);
+}
+
+
+/*
+ * ParameterizationAsString returns the string representation of the set of
+ * rels given in paramRelids.
+ *
+ * Takes a StringInfo so that it doesn't return palloc'd memory. This makes it
+ * easy to call this function as an argument to ereport(), such that it won't
+ * be evaluated unless the message is going to be output somewhere.
+ */
+static char *
+ParameterizationAsString(PlannerInfo *root, Relids paramRelids, StringInfo buf)
```

Comment:
codecov complains about most parts of this function :/ I think we should have some tests for those blocks, where planner_debug_level is set to NOTICE or something >= client_min_messages.

Reply:
Maybe codecov is not right as I see that you already added such tests

Reply:
Oh, I guess this is because, if plan becomes parametrized, then we anyway don't plan the query again (after executing 6? times), so we are never testing the line after 1030.

Reply:
No, you're right, that was not tested before, but it is now.

I think you are getting confused between parameterization for a nested loop join (which is what this patch is mainly about) and evaluating external parameters when executing generic cached plans (which this patch also fixes, but it's pretty minor by comparison).